python-client/giskard/models/base/cache.py

```diff
@@ -0,0 +1,48 @@
+from typing import Optional, Dict, List, Any, Iterable
+
+import numpy as np
+import pandas as pd
+
+NaN = float('NaN')
+
+
+def flatten(xs):
+    for x in xs:
+        if isinstance(x, Iterable) and not isinstance(x, (str, bytes)):
+            yield from flatten(x)
+        else:
+            yield x
+
+
+class ModelCache:
+    prediction_cache: Dict[str, Any]
+
+    vectorized_get_cache_or_na = None
+
+    def __init__(self, cache: Optional[pd.DataFrame] = None):
+        cache = cache if cache is not None else pd.DataFrame(data={}, index=[])
+
+        self.prediction_cache = cache.to_dict()
+
+        self.vectorized_get_cache_or_na = np.vectorize(self.get_cache_or_na, otypes=[object])
```

Comment:
I still believe it is faster to use a `pd.Series` as pointed out here: https://github.com/Giskard-AI/giskard/pull/1002#discussion_r1200479832
It will also make the code more compact and readable.